share/man/man8/fstyp.8
'\" te
.\" Copyright (c) 2003, Sun Microsystems, Inc.  All Rights Reserved
.\" Copyright 1989 AT&T
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH FSTYP 8 "Dec 19, 2007"
.SH NAME
fstyp \- determine file system type
.SH SYNOPSIS
.LP
.nf
\fBfstyp\fR [\fB-a\fR | \fB-v\fR] \fIspecial\fR [:\fIlogical-drive\fR]
.fi

.SH DESCRIPTION
.sp
.LP
\fBfstyp\fR allows the user to determine the file system type of unmounted file
systems using heuristic programs.
.sp
.LP
An \fBfstyp\fR module for each file system type to be checked is executed; each
of these modules applies an appropriate heuristic to determine whether the
supplied \fIspecial\fR file is of the type for which it checks. If it is, the
program prints on standard output the usual file system identifier for that
type (for example, ``ufs'') and exits with a return code of \fB0\fR; if none of
the modules succeed, the error message \fBunknown_fstyp (no matches)\fR is
returned and the exit status is \fB1\fR. If more than one module succeeds, the
error message \fBunknown_fstyp (multiple matches)\fR is returned and the exit
status is \fB2\fR. Other errors are printed in the same format.
.sp
.LP
This command is unreliable and its results should not be used to make any
decisions about subsequent use of a storage device or disk partition.
.SH OPTIONS
.sp
.ne 2
.na
\fB\fB-a\fR\fR
.ad
.RS 6n
Output all available file system attributes. If a file system has been
successfully identified, and this option is specified, the \fIFSType\fR
identifier is followed by one or more "name-value" pairs, one per line, in the
format:
.sp
.in +2
.nf
name: value
.fi
.in -2
.sp

The following conventions are recognized for the file system attributes:
.RS +4
.TP
.ie t \(bu
.el o
String values are put in single quotes.
.RE
.RS +4
.TP
.ie t \(bu
.el o
Nested "name-value" list increases the indentation of its values by four
whitespaces.
.RE
.RS +4
.TP
.ie t \(bu
.el o
For an array of "name-value" pairs, one array entry is printed per line, with
the index following the name in square brackets.
.RE
For instance, in the following example, "\fItop_list\fR" is a "name-value"
list, consisting of a string array "\fIstring_array\fR" and a "name-value" list
array "\fIlist_array\fR". The second "list_array" element is an integer array
"\fIint_array\fR" containing three elements.
.sp
.in +2
.nf
top_string: 'String'
top_list:
    string_array[0]: 'S0'
    string_array[1]: 'S1'
    list_array[0]:
        int_one: 1
        string_two: 'S2'
    list_array[1]:
        int_array[0]: 1
        int_array[1]: 2
        int_array[2]: 3
.fi
.in -2
.sp

In addition to the \fIFSType\fR-specific attributes, the following generic
attributes may be present:
.sp
.ne 2
.na
\fB\fIgen_clean\fR\fR
.ad
.RS 13n
Allowable values are "true" or "false". If the value is "false", the file
system is damaged or was not unmounted cleanly and the \fBfsck\fR(8) command
must be run before this file system can be mounted.
.RE

.sp
.ne 2
.na
\fB\fIgen_guid\fR\fR
.ad
.RS 12n
Globally unique identifier. This string uniquely identifies the file system.
.RE

.sp
.ne 2
.na
\fB\fIgen_version\fR\fR
.ad
.RS 15n
A string describing the file system version.
.RE

.sp
.ne 2
.na
\fB\fIgen_volume_label\fR\fR
.ad
.RS 20n
Volume label, a human-readable string used to either describe or identify the
file system.
.RE

.RE

.sp
.ne 2
.na
\fB\fB-v\fR\fR
.ad
.RS 6n
Produce verbose output. This is usually information about the file systems
superblock and varies across different \fIFSTypes\fR. See \fBufs\fR(7FS),
\fBmkfs_ufs\fR(8), and \fBtunefs\fR(8) for details.
.RE

.SH SEE ALSO
.sp
.LP
\fBfsck\fR(8), \fBmkfs_ufs\fR(8), \fBtunefs\fR(8), \fBattributes\fR(5),
\fBlibfstyp\fR(3LIB), \fBhsfs\fR(7FS), \fBufs\fR(7FS),
\fBpcfs\fR(7FS)
.SH NOTES
.sp
.LP
The use of heuristics implies that the result of\fBfstyp\fR is not guaranteed
to be accurate.
.sp
.LP
This command is unreliable and its results should not be used to make any
decisions about subsequent use of a storage device or disk partition.


usr/src/cmd/fs.d/zfs/fstyp/fstyp.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#pragma ident	"%Z%%M%	%I%	%E% SMI"

/*
 * libfstyp module for zfs
 */
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <libintl.h>
#include <locale.h>
#include <string.h>
#include <libnvpair.h>
#include <libzfs.h>
#include <libfstyp_module.h>
#include <errno.h>

struct fstyp_zfs {
	int		fd;
	nvlist_t	*config;
};

int	fstyp_mod_init(int fd, off_t offset, fstyp_mod_handle_t *handle);
void	fstyp_mod_fini(fstyp_mod_handle_t handle);
int	fstyp_mod_ident(fstyp_mod_handle_t handle);
int	fstyp_mod_get_attr(fstyp_mod_handle_t handle, nvlist_t **attrp);

int
fstyp_mod_init(int fd, off_t offset, fstyp_mod_handle_t *handle)
{
	struct fstyp_zfs *h;

	if (offset != 0) {
		return (FSTYP_ERR_OFFSET);
	}

	if ((h = calloc(1, sizeof (struct fstyp_zfs))) == NULL) {
		return (FSTYP_ERR_NOMEM);
	}
	h->fd = fd;

	*handle = (fstyp_mod_handle_t)h;
	return (0);
}

void
fstyp_mod_fini(fstyp_mod_handle_t handle)
{
	struct fstyp_zfs *h = (struct fstyp_zfs *)handle;

	nvlist_free(h->config);
	free(h);
}

int
fstyp_mod_ident(fstyp_mod_handle_t handle)
{
	struct fstyp_zfs *h = (struct fstyp_zfs *)handle;
	uint64_t state;
	char	*str;
	uint64_t u64;
	char	buf[64];

	if (zpool_read_label(h->fd, &h->config) != 0) {
		return (FSTYP_ERR_NO_MATCH);
	}

	if (nvlist_lookup_uint64(h->config, ZPOOL_CONFIG_POOL_STATE,
	    &state) != 0 || state == POOL_STATE_DESTROYED) {
		nvlist_free(h->config);
		h->config = NULL;
		return (FSTYP_ERR_NO_MATCH);
	}

	/* add generic attributes */
	(void) nvlist_add_boolean_value(h->config, "gen_clean", B_TRUE);
	if (nvlist_lookup_uint64(h->config, "guid", &u64) == 0) {
		(void) snprintf(buf, sizeof (buf), "%llu", (u_longlong_t)u64);
		(void) nvlist_add_string(h->config, "gen_guid", buf);
	}
	if (nvlist_lookup_uint64(h->config, "version", &u64) == 0) {
		(void) snprintf(buf, sizeof (buf), "%llu", (u_longlong_t)u64);
		(void) nvlist_add_string(h->config, "gen_version", buf);
	}
	if (nvlist_lookup_string(h->config, "name", &str) == 0) {
		(void) nvlist_add_string(h->config, "gen_volume_label", str);
	}

	return (0);
}

int
fstyp_mod_get_attr(fstyp_mod_handle_t handle, nvlist_t **attrp)
{
	struct fstyp_zfs *h = (struct fstyp_zfs *)handle;

	*attrp = h->config;
	return (0);
}
